lang/src/accounts/interface.rs

```diff
@@ -0,0 +1,137 @@
+
+/// Type validating that the account is one of a set of given Programs
+///
+/// The `Interface` wraps over the [`Program`](crate::Program), allowing for
+/// multiple possible program ids. Useful for any program that implements an
+/// instruction interface. For example, spl-token and spl-token-2022 both implement
+/// the spl-token interface.
+///
+/// # Table of Contents
+/// - [Basic Functionality](#basic-functionality)
+/// - [Out of the Box Types](#out-of-the-box-types)
+///
+/// # Basic Functionality
+///
+/// Checks:
+///
+/// - `expected_programs.contains(account_info.key)`
+/// - `account_info.executable == true`
+///
+/// # Example
+/// ```ignore
+/// #[program]
+/// mod my_program {
+///     fn set_admin_settings(...){...}
+/// }
+///
+/// #[account]
+/// #[derive(Default)]
+/// pub struct AdminSettings {
+///     ...
+/// }
+///
+/// #[derive(Accounts)]
+/// pub struct SetAdminSettings<'info> {
+///     #[account(mut, seeds = [b"admin"], bump)]
+///     pub admin_settings: Account<'info, AdminSettings>,
+///     #[account(constraint = program.programdata_address()? == Some(program_data.key()))]
+///     pub program: Interface<'info, MyProgram>,
+///     #[account(constraint = program_data.upgrade_authority_address == Some(authority.key()))]
+///     pub program_data: Account<'info, ProgramData>,
+///     pub authority: Signer<'info>,
+/// }
+/// ```
+/// The given program has a function with which the upgrade authority can set admin settings.
+///
+/// The required constraints are as follows:
+///
+/// - `program` is the account of the program itself.
+/// Its constraint checks that `program_data` is the account that contains the program's upgrade authority.
+/// Implicitly, this checks that `program` is a BPFUpgradeable program (`program.programdata_address()?`
+/// will be `None` if it's not).
+/// - `program_data`'s constraint checks that its upgrade authority is the `authority` account.
+/// - Finally, `authority` needs to sign the transaction.
+///
+/// # Out of the Box Types
+///
+/// Between the [`anchor_lang`](https://docs.rs/anchor-lang/latest/anchor_lang) and [`anchor_spl`](https://docs.rs/anchor_spl/latest/anchor_spl) crates,
+/// the following `Interface` types are provided out of the box:
+///
+/// - [`TokenInterface`](https://docs.rs/anchor-spl/latest/anchor_spl/token_interface/struct.TokenInterface.html)
+///
+#[derive(Clone)]
+pub struct Interface<'info, T>(Program<'info, T>);
+impl<'a, T: CheckId> TryFrom<&AccountInfo<'a>> for Interface<'a, T> {
+    type Error = Error;
+    /// Deserializes the given `info` into a `Program`.
+    fn try_from(info: &AccountInfo<'a>) -> Result<Self> {
+        T::check_id(info.key)?;
+        if !info.executable {
+            return Err(ErrorCode::InvalidProgramExecutable.into());
+        }
+
+        Ok(Self(Program::new(info.clone())))
+    }
+}
+impl<'info, T> Deref for Interface<'info, T> {
+    type Target = Program<'info, T>;
+    fn deref(&self) -> &Self::Target {
```

Comment:
dumb rust question about refs. 

Interface derefs to `Program<'info, T>`, while `Program` itself derefs into the `.info`. When (if ever) would that second deref occur? Whereas, in InterfaceAccount there's a double deref, e.g. `&self.o.deref()`. Why the difference?

Reply:
It's a good question! Depending on how you use the struct, on or both derefs can happen. Here's an example in `token_proxy`:

```
    pub fn proxy_transfer(ctx: Context<ProxyTransfer>, amount: u64) -> Result<()> {
        let account_info: &AccountInfo<'_> = &ctx.accounts.token_program;
        msg!("{}", account_info.owner);
        let program: &Program<'_, TokenInterface> = &ctx.accounts.token_program;
        msg!("{:?}", program.programdata_address()?);
        #[allow(deprecated)]
        token_interface::transfer(ctx.accounts.into(), amount)
    }
```

`token_program` is `Interface<TokenInterface>` but I can immediately "downcast" it as either `AccountInfo` thanks to the double `deref` or to a `Program` thanks to the single.  So it's actually quite useful!

Reply:
Ah yes, I see what you mean now, good catch. I'm leaning towards having `InterfaceAccount` deref to `Account` to reduce the amount of code, and also keep things consistent. What do you think?

Reply:
I prefer derefing straight to the underlying type if Account/Interface, and account info if Program/Interface. Unless there's something on `Program`/`Account` that is really important for people to get access to. Just because I think `&mut *ctx.accounts.stuff` is a much more common pattern since you want to access the underlying `Stuff` type. 

Plus I think you covered all of the impls on `Account` in `Interface` so there shouldn't be anything missing?

Reply:
Works for me, I'll do that!

Reply:
Changed the interface impl in [1c55bb2](https://github.com/coral-xyz/anchor/pull/2386/commits/1c55bb222422002ef577204ff8fe3aa93948ec6f)